### PR TITLE
Adds hierarchy functionality to list_blobs mentioned in issue #55

### PR DESCRIPTION
**Pull request by cantpitch**
Here is an implementation of the by_hierarchy functionality in issue #55.

### CODE REVIEW

File: R/blob_client_funcs.R [hunk 1]
```diff
@@ -350,11 +382,11 @@ list_blobs <- function(container, dir="/", info=c("partial", "name", "all"),
             namecol <- which(ndf == "Name")
             sizecol <- which(ndf == "Content-Length")
             names(df)[c(namecol, sizecol)] <- c("name", "size")
-            df$size <- as.numeric(df$size)
+            df$size <- if (!is.null(df$size)) as.numeric(df$size) else NA
 
             # needed when dir was created using ADLS API
             # this works because content-type is always set for an actual file
-            df$isdir <- is.na(df$`Content-Type`)
+            df$isdir <- if (!is.null(df$`Content-Type`)) is.na(df$`Content-Type`) else T
```
Comment (hongooi73) [inline R/blob_client_funcs.R:389]: Use `TRUE`, not `T`

File: R/blob_client_funcs.R [hunk 2]
```diff
@@ -339,8 +354,25 @@ list_blobs <- function(container, dir="/", info=c("partial", "name", "all"),
...
+        df_prefixes <- do.call(rbind, prefix_rows)
+        df_blobs <- do.call(rbind, blob_rows)
+        
+        if (is.null(df_prefixes) & is.null(df_blobs)) 
+            return(data.frame())
+        else if (is.null(df_prefixes))
+            df <- df_blobs
+        else if (is.null(df_blobs))
+            df <- df_prefixes
+        else {
```
Comment (hongooi73) [inline R/blob_client_funcs.R:368]: Opening brace should always be on a new line.

**Comment by hongooi73** [comment]
I've now had a chance to take a closer look at this, and I see that the `list_blobs` call only returns the first level of the hierarchy. The .NET implementation that you linked before actually call the API endpoint recursively, to enumerate all levels. Do you think it would be worthwhile to do that as well in `list_blobs`? That seems to be more useful than just returning one level, without any hint of how to get the other levels.

**Comment by cantpitch** [comment]
You can use this recursively to fetch the hierarchy. If you assume blob
container that's structured like:

toplevel/
    blob1.txt
    level1/
        blob2.txt
        level2/
            blob3.txt

then
   list_blobs(container, prefix = "toplevel/", by_hierarchy = T)

will return
    toplevel/blob1.txt
    toplevel/level1/

and then
   list_blobs(container, prefix = "level1/", by_hierarchy = T)

will return
   toplevel/level1/blob2.txt
   toplevel/level2/

So you can recursively call it on the prefixes for BlobPrefixes.

Make sense? We can add some documentation to make it clear how to use it.

**Comment by hongooi73** [comment]
Yes, I think it would be good to have this built into `list_blobs`, rather than the user having to manually list each level/directory. Unless you see a good reason for not doing this?

**Comment by cantpitch** [comment]
Well, if you just want the entire hierarchy you can call list_blobs with
by_hierarchy = FALSE (the default) and will return everything under the
prefix you specified. The advantage of the by_hierarchy flag is that it
*only* gets the next level. The use case is for when you have a huge
hierarchy and you want to only get one level, do some processing, get the
next level, etc. So the recursive aspect is the advantage for the user.
I've got a storage account with about 3.5MM blobs (totalling about 3 TB)
that are in a date hierarchy like:

storage_account/2019/05/24/123455/file.json

so I _don't_ want it to pull in the entire hierarchy in one call, I want to
iterate through the hierarchy recursively. If the user doesn't want to list
each level/directory, they can just call without by_hierarchy and it will
do the original behavior.

**Comment by hongooi73** [comment]
Ok, that makes sense. Can you put something in the documentation describing this use case and what `by_hierarchy=TRUE` returns, and we can merge this.

**Comment by hongooi73** [comment]
Actually, I've just realised that the functionality here is basically the same as that provided by the `recursive` argument for the file and ADLS2 storage types. That is, `list_azure_files(*, recursive=FALSE)` and `list_adls_files(*, recursive=FALSE)` will both give you the result for one level of the hierarchy.

Rather than having an extra `by_hierarchy` argument, can you reuse the existing `recursive`?

**Comment by hongooi73** [comment]
Actually, let me handle this. Looks like there's some annoying warts related to hierarchical namespaces.

**Comment by hongooi73** [comment]
So, the blob-specific requirement to append a "/" to get a subdirectory listing doesn't play well with `multidownload_blob` (which calls `list_blobs` to see what files to download). It's also inconsistent with the behaviour for file and ADLS2 storage, which ignore a trailing "/". Eg, `list_adls_files(dir="dirname")` and `list_adls_files(dir="dirname/")` return identical results.

I'll modify `list_blobs` so that it behaves this way as well, ie, you won't need to remember to add a trailing "/". I don't think this should cause any problems; if this doesn't work, let me know.

**Pull request merged**